### ISSUE TITLE
tf.nn.softmax overflow due to exp(x)

### ISSUE THREAD

**Issue by weiliu620** [tensorflow/tensorflow#10142]
From reading the doc and code of ```tf.nn.softmax```, it looks like this function naively call a ```_softmax``` function which may have the overflow problem if the x in exp(x) is too big. (underflow can happen, too)

This is a typical problem of machine learning, and googling 'log sum exp' gives some information. TF even has a function ```reduce_logsumexp``` to calculate the value in a robust way. 

I think it makes sense that ```nn.softmax``` also use similar method of ```reduce_logsumexp```. I ran into a loss  = NaN today and suspect this is the cause. 

Thanks,

**Comment by aselle** [comment]
I'm not so sure you can do the exact same thing. That method is taking into account that if you scale the computation based on the range of the existing values, you can exp and reduce (sum) in a space where there is more floating accuracy. You'd have to do that scaling and return the caling to the user. That would make for a complicated interface for users of tf.softmax.  As you say this is a known problem, which is why we have reduce_logsumexp (and softmax_cross_entropy_with_logit), but tf.nn.softmax may be useful in other situations.

**Comment by ekelsen** [comment]
note that softmax does subtract the maximum logit, so the max value it will exponentiate will be 0.  I don't think this is the cause of your NaN.  Without more information it's hard to say.

**Comment by tensorflowbutler** [comment]
This issue is automatically closed due to lack of activity. Please re-open if this is still an issue for you. Thanks!

**Comment by RT-TL** [comment]
@ekelsen I read this statement also regarding other softmax functions such as tf.nn.sparse_softmax_cross_entropy_with_logits. However, I do have this issue. Using TFdebug I figured out that an nan value (inf) often results from the output of the softmax calculation done within sparse_softmax_... Shouldn't this be impossible using a max value subtraction?

**Comment by jesseengel** [comment]
Another source of NaNs in sparse_softmax can be from passing in an int label that is >= than the number of dimensions in the logits. Hope that helps.
